packages/ui-react-utils/src/getElementType.ts

```diff
@@ -37,15 +47,14 @@ import { ComponentType, ReactElement } from 'react'
  * @param {Function} getDefault an optional function that returns the default element type
  * @returns {String} the element type
  */
-function getElementType<
-  T extends { as?: ReactElement | string; [key: string]: any }
->(
-  // TODO remove |any and as any casts when propTypes are removed
+function getElementType<T extends ComponentWithAsProp>(
+  // TODO: remove |any and as any casts when propTypes are removed
   Component: ComponentType<T> | any,
   props: T,
-  getDefault?: () => string | ReactElement
-) {
-  if (props.as && props.as !== (Component.defaultProps as any).as) {
+  getDefault?: () => AsElementType
+  // TODO: add better typing to the return value
+): any {
```

Comment:
It is really hard to type this one, probably needs a bigger refact. I'll open a ticket for it.